sycl/doc/extensions/EnqueueBarrier/enqueue_barrier.asciidoc

```diff
@@ -0,0 +1,315 @@
+:lang: en
+
+:blank: pass:[ +]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+== Introduction
+IMPORTANT: This specification is a draft.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by permission by Khronos.
+
+NOTE: This document is better viewed when rendered as html with asciidoctor.  GitHub does not render image icons.
+
+This document presents a series of changes proposed for a future version of the SYCL Specification. The goal of this proposal is to provide non-blocking APIs that provide synchronization on SYCL command queue for programmers.
+
+
+== Name Strings
+
++SYCL_INTEL_enqueue_barrier+
+
+== Notice
+
+Copyright (c) 2019-2020 Intel Corporation.  All rights reserved.
+
+== Status
+
+Final Draft
+
+This is a preview extension specification, intended to provide early access to a feature for review and community feedback. When the feature matures, this specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are subject to change they are not intended to be used by shipping software products.
+
+== Version
+
+Built On: {docdate} +
+Revision: 1
+
+== Contact
+Please open an issue in the https://github.com/intel/llvm/tree/sycl/sycl/doc/extensions/[extensions repository]
+
+== Dependencies
+
+This extension is written against the SYCL 1.2.1 specification, Revision v1.2.1-6.
+
+== Overview
+
+SYCL 1.2.1 defines a graph-based task execution model, based on kernels or explicit memory operations submitted to out-of-order queues. Dependencies between these kernels are represented by
+accessors that form data dependence edges in the execution graph. The USM extension <<usmlink,[1]>> doesn't have accessors, so instead solves
+this by defining `handler::depends_on` methods to specify event-based control dependencies between command groups.
+
+There are situations where defining dependencies based on events is more explicit than desired or required by an application. For instance, the user may know that a given task depends on all previously submitted tasks. Instead of explicitly adding all the required depends_on calls, the user could express this intent via a single call, making the program more concise and explicit.
+
+To simplify the interface, this extension adds an enqueued barrier that provides synchronization on SYCL command
+queues, with the following properties:
+
+1. Enqueued barriers are non-blocking from the host perspective.  The barrier is enqueued, so operates as part of the execution graph asynchronously from host program execution
+2. Command groups submitted to the same queue after the barrier is enqueued are not scheduled for execution until the barrier wait conditions are satisfied
```

Comment:
Would it be better do describe this using terms from the SYCL spec?

For example, `scheduled` is kind of a new term, I cannot find it in Section 3.4.1.2 SYCL Command Groups and Execution Order or in Glossary

Reply:
"scheduled" is not a new term, please refer to Section 4.6.5 Table 4.22 Member functions of class queue, page 75 of SYCL spec, when describing queue::submit():

Submit a command group function object to the queue, in order to be scheduled for execution on the device.

Reply:
Missed that one, thanks for pointing to it